packages/host/app/resources/search.ts

```diff
@@ -1,79 +1,189 @@
+  private currentResults = new TrackedMap<string, CardResource>();
 
   modify(_positional: never[], named: Args['named']) {
-    let { query, realms } = named;
+    let { query, realms, isLive, doWhileRefreshing } = named;
     this.realmsToSearch = realms ?? this.realmServer.availableRealmURLs;
 
     this.loaded = this.search.perform(query);
     waitForPromise(this.loaded);
+
+    if (isLive) {
+      this.subscriptions = this.realmsToSearch.map((realm) => ({
+        url: `${realm}_message`,
+        unsubscribe: subscribeToRealm(
+          `${realm}_message`,
+          ({ type, data }: { type: string; data: string }) => {
+            let eventData = JSON.parse(data);
+            // we are only interested in incremental index events
+            if (type !== 'index' || eventData.type !== 'incremental') {
+              return;
+            }
+            this.search.perform(query);
+            if (doWhileRefreshing) {
+              this.doWhileRefreshing.perform(doWhileRefreshing);
+            }
+          },
+        ),
+      }));
+
+      registerDestructor(this, () => {
+        for (let subscription of this.subscriptions) {
+          subscription.unsubscribe();
+        }
+      });
+    }
   }
 
+  @cached
   get instances() {
-    return this.isLoading ? this.staleInstances : this._instances;
+    return [...this.currentResults.values()].map((r) => r.card) as CardDef[]; // empty results are filtered out already
   }
 
+  @cached
   get instancesByRealm() {
-    return this.isLoading ? this.staleInstancesByRealm : this._instancesByRealm;
+    return this.realmsToSearch
+      .map((realm) => {
+        let cards = this.instances.filter((card) => card.id.startsWith(realm));
+        return { realm, cards };
+      })
+      .filter((r) => r.cards.length > 0);
+  }
+
+  private doWhileRefreshing = restartableTask(
+    async (
+      doWhileRefreshing: (ready: Promise<void> | undefined) => Promise<void>,
+    ) => {
+      await doWhileRefreshing(this.loaded);
+    },
+  );
+
+  // By default, this does a tracked read from seenCards so that your
+  // answer can be invalidated if a new card is discovered. Internally, we also
+  // use it untracked to implement the read-through cache.
+  private seenCard(url: string, tracked = true): CardResource | undefined {
+    let resource = this.cardResources.get(url);
+    if (resource) {
+      return resource;
+    }
+    if (tracked) {
+      this.seenCardResource.has(url);
+    }
+    return undefined;
+  }
+
+  private getOrCreateCardResource(
+    urlOrDoc: string | SingleCardDocument,
+  ): CardResource {
+    let url = asURL(urlOrDoc);
+    // this should be the only place we do the untracked read. It needs to be
+    // untracked so our `this.cardResources.set` below will not be an assertion.
+    let resource = this.seenCard(url, false);
+    if (!resource) {
+      resource = getCard(this, () => urlOrDoc, {
+        isLive: () => true,
```

Comment:
I'm assuming that all cards we get back from search (whether the search is live or not) are live cards. we can easily amend this if that is a bad assumption.